File: LeetCode/Simulation/Golang/Leetcode592-fraction-addition-and-subtraction.go
package sbgo

import (
	"fmt"
	"unicode"
)

func fractionAddition(expression string) string {
	n, i, up, down := len(expression), 0, 0, 1
	for i < n {
		sign, up1, down1 := 1, 0, 0
		if expression[i] == '-' || expression[i] == '+' {
			if expression[i] == '-' {
				sign = -1
			}
			i++
		}
		for i < n && unicode.IsDigit(rune(expression[i])) {
			up1 = up1*10 + int(expression[i]-'0')
			i++
		}
		i++
		up1 = sign * up1

		for i < n && unicode.IsDigit(rune(expression[i])) {
			down1 = down1*10 + int(expression[i]-'0')
			i++
		}
		up = up1*down + up*down1
		down *= down1
	}
	if up == 0 {
		return "0/1"
	}
	g := gcd(abs(up), down)
	return fmt.Sprintf("%d/%d", up/g, down/g)
}

func gcd(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd(b, a%b)
}

func abs(i int) int {
	if i < 0 {
		return -i
	}
	return i
}

File: LeetCode/Simulation/Python/Leetcode592-fraction-addition-and-subtraction.py
class Solution:
    def fractionAddition(self, expression: str) -> str:
        n,i,up,down = len(expression),0,0,1

        while i < n:
            sign,up1,down1 = 1,0,0
            if expression[i] == '-' or expression[i] == '+':
                sign = -1 if expression[i] == '-' else 1
                i += 1
            while i < n and expression[i].isdigit():
                up1 = up1 * 10 + ord(expression[i]) - ord('0') 
                i += 1
            up1 = sign * up1

            i += 1
            while i < n and expression[i].isdigit():
                down1 = down1 * 10 + ord(expression[i]) - ord('0')
                i += 1
            up = up1 * down + up * down1
            down *= down1
        if up == 0: return '0/1'
        g = gcd(abs(up),down)
        return "{}/{}".format(up // g,down // g)            
